File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> list = new ArrayList<List<Integer>>();
        if(root == null) return list;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        
        while(queue.size() > 0){
            int size = queue.size();
            List<Integer> currentList = new ArrayList<Integer>();
            
            for(int i = 0; i < size; i++){
                TreeNode current = queue.poll();
                currentList.add(current.val);
                if(current.left != null) queue.add(current.left);
                if(current.right != null) queue.add(current.right);
            }
            list.add(currentList);
        }
        
        return list;
    }
}

File: binary-tree-level-order-traversal/binary-tree-level-order-traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        
        val outerList = ArrayList<ArrayList<Int>>()
        if(root == null) return outerList
        
        
        val queue = LinkedList<TreeNode>()
        
        queue.offer(root)
        
        while(queue.size > 0){
            val size = queue.size
            val innerList = ArrayList<Int>()
            
            for(i in 0 until size){
               val node = queue.poll() 
               node?.let{
                   innerList.add(it.`val`)
                   it.left?.let{ left ->
                       queue.offer(left)
                   }
                   it.right?.let{ right ->
                       queue.offer(right)
                   }
               }
            }
            
            outerList.add(innerList)
            
        }
        return outerList
        

        
    }
}